ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/EthFeeHistory.java

```diff
@@ -143,14 +150,31 @@ public JsonRpcResponse response(final JsonRpcRequestContext request) {
     final Optional<List<List<Wei>>> maybeRewards =
         maybeRewardPercentiles.map(
             rewardPercentiles ->
-                LongStream.range(oldestBlock, lastBlock)
-                    .mapToObj(blockchain::getBlockByNumber)
-                    .flatMap(Optional::stream)
+                blockHeaders.stream()
+                    .parallel()
                     .map(
-                        block ->
-                            computeRewards(
-                                rewardPercentiles.stream().sorted().collect(toUnmodifiableList()),
-                                block))
+                        blockHeader -> {
+                          final RewardCacheKey key =
+                              new RewardCacheKey(blockHeader.getBlockHash(), rewardPercentiles);
+                          return Optional.ofNullable(cache.getIfPresent(key))
+                              .or(
+                                  () -> {
+                                    Optional<Block> block =
+                                        blockchain.getBlockByHash(blockHeader.getBlockHash());
+                                    return block.map(
+                                        b -> {
+                                          List<Wei> rewards =
+                                              computeRewards(
+                                                  rewardPercentiles.stream()
+                                                      .sorted()
+                                                      .collect(toUnmodifiableList()),
+                                                  b);
+                                          cache.put(key, rewards);
+                                          return rewards;
+                                        });
+                                  });
+                        })
+                    .flatMap(Optional::stream)
                     .collect(toUnmodifiableList()));
```

Comment:
credit to chatGPT for noticing that this sort can be done outside of blockHeaders.stream(), e.g.:

```suggestion
    final Optional<List<List<Wei>>> maybeRewards =
        maybeRewardPercentiles.map(
            rewardPercentiles -> {
              var sortedPercentiles = rewardPercentiles.stream().sorted().toList();
              return blockHeaders.stream()
                  .parallel()
                  .map(
                      blockHeader -> {
                        final RewardCacheKey key =
                            new RewardCacheKey(blockHeader.getBlockHash(), rewardPercentiles);
                        return Optional.ofNullable(cache.getIfPresent(key))
                            .or(
                                () -> {
                                  Optional<Block> block =
                                      blockchain.getBlockByHash(blockHeader.getBlockHash());
                                  return block.map(
                                      b -> {
                                        List<Wei> rewards = computeRewards(sortedPercentiles, b);
                                        cache.put(key, rewards);
                                        return rewards;
                                      });
                                });
                      })
                  .flatMap(Optional::stream)
                  .toList();
            });
```

Reply:
I will definitely check the performances and do another PR if needed